### DISCUSSION TITLE
0073 XLS-73d: AMMClawback

### DISCUSSION THREAD

**Discussion by yinyiqian1** [XRPLF/XRPL-Standards#212] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by ximinez** [comment]
The way the new `AMMClawback` transaction will work is pretty neat. If I'm understanding correctly, it's effectively simulating two transactions atomically: 
1. Simulating as if the `Holder` account submitted an `AMMWithdraw` with the `tfWithdrawAll` flag.
2. Issuing a `Clawback` from the issuer with the amount set to the entire amount of the issuer's token withdrawn by step 1.

Except it will (presumably) avoid rounding issues.

Is this roughly an accurate description?

Two questions:
1. Why not include an `Amount` field (maybe optional), like the one in `Clawback` so that the issuer can only take back a portion of the holdings?
2. If the issuer issued both of the tokens in the AMM, do they get all of both tokens withdrawn by the `AMMClawback`?

**Reply by yinyiqian1** [reply]
1.  I think it is a good idea to add it to make it more flexible. I'll add it to the spec.
2. By adding the amount, the issuer is able to specify which token to clawback. If the issuer issued both of the tokens in the AMM, then the token the issuer specifies in the `Amount` will be back to the issuer. The other token will go back to the holder.

**Reply by ximinez** [reply]
2. Idea: Add an `Amount` and `Amount2`. Make both optional. If `Amount2` is included, `Amount` must be included, and they must be for the different currencies, and the issuer must have issued both. If neither are included, the issuer gets back the full amount of all tokens they issued.

**Reply by gregtatcam** [reply]
Note that the Withdraw has to be proportional. Might be better specifying the amount of LPToken to withdraw instead.

**Reply by yinyiqian1** [reply]
> 2. Idea: Add an `Amount` and `Amount2`. Make both optional. If `Amount2` is included, `Amount` must be included, and they must be for the different currencies, and the issuer must have issued both. If neither are included, the issuer gets back the full amount of all tokens they issued.

Because the withdrawal is proportional, it's better to only specify one asset, and giving back the other asset. If the issuer wants to claw back both tokens, he can call the regular Clawback after the AMMClawback transaction to get back the other token

**Reply by ximinez** [reply]
> Note that the Withdraw has to be proportional. Might be better specifying the amount of LPToken to withdraw instead.

Great suggestion!

> he can call the regular Clawback after the AMMClawback transaction to get back the other token

That could introduce a race condition if the `Holder` is watching for something like this.

**Reply by shawnxie999** [reply]
@ximinez 

> That could introduce a race condition if the Holder is watching for something like this.

could you elaborate on this?

**Reply by yinyiqian1** [reply]
> Note that the Withdraw has to be proportional. Might be better specifying the amount of LPToken to withdraw instead.

Yes, we can make the request like:

{
  "TransactionType": "AMMClawback",
  "Account": "rPdYxU9dNkbzC5Y2h4jLbVJ3rMRrk7WVRL",
  "Holder": "rvYAfWj5gh67oV6fW32ZzP3Aw4Eubs59B",
  "AMMAccount": "rp2MaZMQDpgAHwWbADaQMrmf4AD5JsPQUR",
  "Amount":  {
          "currency" : "0348ACDAB7D138D5D2AAD95492FEB7ECD5D55CC1",
          "issuer" : "rPdYxU9dNkbzC5Y2h4jLbVJ3rMRrk7WVRL",
          "value" : "1000"
      }
  "Flags": 1,
  "Fee": 10
}

Amount is optional, if not provided, we will just clawback all the LPtokens.
subfield currency is LPtoken's hash, which should match with the AMMAccount.
subfield issuer should match with issuer, which is the Account field

**Reply by gregtatcam** [reply]
I think `issuer` has to match AMMAccount in this case, right?

**Reply by yinyiqian1** [reply]
> I think `issuer` has to match AMMAccount in this case, right?

yes make sense. It should be AMMAccount

**Reply by ximinez** [reply]
> > That could introduce a race condition if the Holder is watching for something like this.
> 
> could you elaborate on this?

I don't think there's anything in `AMMClawback` / `Clawback` that requires the trust line to be frozen. If I'm wrong about this, my whole point is moot. So imagine the scenario:
1. Holder holds two tokens issued by Issuer and puts them into an AMM using those two tokens.
2. Holder knows Issuer is upset with them going to be doing something, so watches the ledger.
3. Issuer neglects to freeze the trust lines, and sends an `AMMClawback` for one of the tokens.
4. At about the same time, Issuer sends a regular `Clawback` for the other token.
5. Holder sees the `AMMClawback`, and quickly sends a `Payment` transaction to send the other token *somewhere else*.
6. Now the `Clawback` and `Payment` are in a race to see which gets processed first. Chances are the `Clawback` will win because of the canonical transaction ordering, but it's not guaranteed if Issuer's two transactions are in two different ledgers.

Put another way, for the Issuer to guarantee their clawback would work, they'd have to send a bunch of transactions:
1. Freeze one trust line
2. Freeze the other trust line
3. AMMClawback
4. Regular Clawback
5. Unfreeze one trust line
6. Unfreeze the other trust line.

That's why I think requiring `Amount` to be the LPToken solves all these problems. It removes all the ambiguity, and rolls those 6 transactions into 1.

**Reply by gregtatcam** [reply]
I think `Amount` can still be used for one token and `Amount` of another token can be calculated proportionally to the pool balances. Then can use equalWithdrawLimit() with both amounts specified. This is still going to be atomic clawback.

**Reply by yinyiqian1** [reply]
The situation that an issuer issues both assets is rare.
I'm hesitant to include both Amount and Amount2 as it may cause confusion among most issuers, who typically only want to claw the asset they issued and do not care about the other asset. Although Amount and Amount2 are optional fields, their presence may lead to unnecessary complexity.
Issuers who have issued both assets can safely clawback both, provided they follow best practices and freeze the trustlines initially. This scenario is expected to be rare. And we can mention it in the doc.
And issuers mostly prefer to specify the token they issued instead of the lptokens.

**Comment by mvadari** [comment]
> Prohibit a wallet from depositing new tokens (single-sided and double-sided) into an AMM pool if at least one of the tokens the wallet owns has been frozen (either individually or globally) by the token issuer

Is withdrawing already prohibited? If not, is there a reason why it is not being prohibited in this proposal?

**Reply by gregtatcam** [reply]
> Is withdrawing already prohibited? If not, is there a reason why it is not being prohibited in this proposal?

Withdrawing of a frozen token is prohibited.

**Reply by Laszlo23** [reply]
So can you withdrawle a frozen token? This is still a little confusing for me

**Reply by joynancy** [reply]
Currently, without this proposed amendment, users cannot withdraw from the AMMPool if the pool contains a pair involving the user's frozen asset.

**Comment by cindyyan317** [comment]
>AMMCreate is allowed for clawback-enabled issuer

For such pool, do we need to include a boolean **allowClawback** in **amm_info's** response? eg. account_info also includes **allowTrustLineClawback**

**Reply by yinyiqian1** [reply]
Yes. We can add it later

**Reply by yinyiqian1** [reply]
This is still on the account level. So we will not add it on the amm level.

**Comment by DeFiDev589** [comment]
Is the clawback feature able to be used to prohibit an account that is not on a blacklist but just trying to single deposit Asset “A” which is an asset that anyone can acquire on the XRP Ledger (For example: XRP or RLUSD) into an AMM Pool paired to Asset “B” which is a regulated asset that they are not authorized to acquire or hold?
It’s my understanding that prior to this amendment anyone could single deposit and single withdraw just ASSET “A” without ever directly interacting with 
ASSET “B” but the liquidity provider is able to gain exposure to 50:50 of both assets in the Pool expressed through the LP Token which accrues the AMM fees, CAM revenue, Harvesting Volatility, but also any other benefits that are inherently embedded in ASSET “B” 
I hope this new amendment doesn’t remove this accessibility. My understanding is liquidity provider who single deposits only ASSET “A” (XRP or RLUSD)  

 In a hypothetical AMM Pool between RLUSD/TBILL a retail user could actually single deposit RLUSD and gain exposure to 50% of $TBILL which accrues the off-chain T-Bill Yield in its value gradually increasing.

This makes the LP Token extremely powerful because It’s effectively a Super Stablecoin that is accruing multi Yield Streams that include the On-Chain AMM FEES, CAM revenue , Harvesting Volatility BUT with the addition of OFF-CHAIN Yield from short term US Treasury T-Bills on 50% of the value LP Tokens which is their ownership share.  WoW 🤩These hypothetical LP Tokens are Ultra Pristine COLLATERAL. Perfect to  use as collateral to borrower $RLUSD or $XRP creating borrow side demand in a secured DeFi Money Market Protocol on EVM Sidechain. 

I appreciate your response as it will help the community understand what the possible potential opportunities w we might see built out by Dapp Devs on EVM sidechain utilizing the AMM LP Tokens or in Future on XRPL once supporting amendments are passed and implemented. 
Unique advantage is the guaranteed redemption by the XRP Ledger which could trigger liquidation, which I think is extremely unlikely and would only occur if there was a complete collapse of either asset or critical security bug. 

Thank You to all the brilliant minds who have been rigorously working continuously enhancing and maintaining the XRPL. Including all those individuals and businesses and universities and projects building on XRPL that run critical infrastructure including nodes, validators, and others. The community XRP Ledger Devs, Cryptographers, Mathematicians, Researchers, and Engineers that are executing on Ripple’s brilliant and very ambitious Roadmap for creating the foundation to support RWA Tokenization, Compliance, Identity Verification, and new enhancements to the core ledger or supporting the necessary 
infrastructure to support TradFi adoption institutional DeFi Roadmap. 

Final question, can the current XLS-30d AMM SPEC be used with a Stablecoin like RLUSD and an Asset like TBILL tokens that are based on ERC-4626 Vault Contract that gradually but slowly increases the price from the Vault contract distribution of the Yield directly into the the TBILL Token market price, 
Thank You

-Mickey B. Fresh “The DeFi Standard"

**Reply by yinyiqian1** [reply]
> Is the clawback feature able to be used to prohibit an account that is not on a blacklist but just trying to single deposit Asset “A” which is an asset that anyone can acquire on the XRP Ledger (For example: XRP or RLUSD) into an AMM Pool paired to Asset “B” which is a regulated asset that they are not authorized to acquire or hold?

If a holder is frozen for asset A, he can not withdraw either A or B(which is the pair asset with A in the AMM pool). Imagine if the holder asset A is frozen for a while and unfrozen later, we don't want him to do some unnecessary things like depositing B into the pool and get any potential loss because he's not able to withdraw anyway.

> Final question, can the current XLS-30d AMM SPEC be used with a Stablecoin like RLUSD and an Asset like TBILL tokens that are based on ERC-4626 Vault Contract that gradually but slowly increases the price from the Vault contract distribution of the Yield directly into the the TBILL Token market price

TBILL is based on MPToken. And MPToken interacting with AMM pools is on the roadmap.

**Reply by DeFiDev589** [reply]
> > Is the clawback feature able to be used to prohibit an account that is not on a blacklist but just trying to single deposit Asset “A” which is an asset that anyone can acquire on the XRP Ledger (For example: XRP or RLUSD) into an AMM Pool paired to Asset “B” which is a regulated asset that they are not authorized to acquire or hold?
> 
> If a holder is frozen for asset A, he can not withdraw either A or B(which is the pair asset with A in the AMM pool). Imagine if the holder asset A is frozen for a while and unfrozen later, we don't want him to do some unnecessary things like depositing B into the pool and get any potential loss because he's not able to withdraw anyway.
> 
> > Final question, can the current XLS-30d AMM SPEC be used with a Stablecoin like RLUSD and an Asset like TBILL tokens that are based on ERC-4626 Vault Contract that gradually but slowly increases the price from the Vault contract distribution of the Yield directly into the the TBILL Token market price
> 
> TBILL is based on MPToken. And MPToken interacting with AMM pools is on the roadmap.

THANK YOU 🙏 for your response 
It does make a lot of sense that the “TBILL token”  will be based on MPToken spec. After watching the presentation on MPTokens from APEX 2024, I’m so excited for new possibilities MPTokens will bring to the XRP Ledger and the XRPL ecosystem in the future. Especially, with tokenized RWA’s . 
Keep up the great work and looking forward to hearing more about this in the future. Thank You!

-Mickey B. Fresh “The DeFi Standard”

**Comment by gregtatcam** [comment]
Why are we only blocking offer where `TakerGets` has a frozen `LPToken`? `TakerPays` also enables to transfer `LPToken` into the owner's account.

**Reply by yinyiqian1** [reply]
The freeze lptoken part will be moved to another amendment.

**Comment by gregtatcam** [comment]
Providing `Asset` in `AMMClawback` is redundant if `Amount` is also included.  Why not only include `Amount` only? If an issuer wants to clawback everything then `Amount` is 0.

**Reply by yinyiqian1** [reply]
updated to use `Asset` and `Asset2` to identify the AMM instance.

**Comment by gregtatcam** [comment]
`AMMClawback` uses `AMMAccount` to identify AMM instance. All AMM transactions (except for `AMMCreate`) identify AMM by the token pair. The general consensus was that it's simpler for developers. Why are we changing to `AMMAccount` and making it inconsistent with other AMM transactions? Now developers have to maintain both the token pair and the account. Let's replace `AMMAccount` with `Asset2`. Then `Asset` and `Asset2` identify AMM instance. `Asset` also identifies the token being claw backed. `Amount` then can be omitted if an issuer wants to claw back all tokens.

**Reply by yinyiqian1** [reply]
updated to use `Asset` and `Asset2`

**Comment by yinyiqian1** [comment]
Per the new [XLS Contributing process](https://github.com/XRPLF/XRPL-Standards/blob/master/CONTRIBUTING.md), I propose that we can move this discussion to a draft spec: https://github.com/XRPLF/XRPL-Standards/pull/233

Please comment here if you would like to object to moving this spec/discussion forward in the process into a DRAFT spec.

**Comment by nhartner** [comment]
Related to a previous question, what's the behavior of AMMClawback transaction against a Holder in an AMM pool where the Holder is only authorized to hold one of the assets from 2 different issuers?
Scenario:
- Token 1: FOO - issued by FOO that has no restrictions
- Token 2: BAR - issued by BAR that has authorized trustlines enabled. 
- HOLDER owns 100 FOO tokens
- HOLDER does not have an authorized trustline to issuer BAR so cannot hold BAR tokens
- HOLDER deposits 100 FOO into the FOO/BAR AMM pool. HOLDER receives LP tokens representing 10% of the pool which now has 500 FOO and 500 BAR.
- Issuer FOO submits an AMMClawback transaction against HOLDER and the FOO/BAR AMM pool

What is the result of the AMMClawback transaction? Under normal circumstances, I understand that 50 FOO would be transferred from the AMM pool to Issuer FOO and 50 BAR would be transferred to HOLDER. But in this scenario, HOLDER is not authorized to hold BAR. As such, I can see a few possible options:
- AMMClawback transaction is rejected. Issuer FOO will not be able to clawback tokens from this HOLDER in the pool.
- 100 FOO are transferred from AMM to Issuer FOO. No BAR tokens are removed from pool.

**Reply by nhartner** [reply]
Even without authorized trustlines, if HOLDER does a single-sided deposit into a pool, they may not have a trustline for both assets in the pool. This could result in the AMMClawback attempting to return a token to the HOLDER for which they do not have a trustline.

**Reply by yinyiqian1** [reply]
We may need to add another AMM restriction in this amendment that an account can not deposit into the AMM pool if the account is not authorized to hold either token. I'll add it in to the spec once confirmed.

Currently if a user only have a trustline with one token, the user can still swap out the other token that the user does not have trustline with, is that correct? @gregtatcam

**Reply by DeFiDev589** [reply]
> We may need to add another AMM restriction in this amendment that an account can not deposit into the AMM pool if the account is not authorized to hold either token. I'll add it in to the spec once confirmed.
> 
> Currently if a user only have a trustline with one token, the user can still swap out the other token that the user does not have trustline with, is that correct? @gregtatcam


I think it should be considered what the implications and consequences are of FORCING the liquidity provider to establish a Trustline with the issuer of the other asset in an AMM POOL when user only wants to deposit XRP and redeem XRP?
I understand Ripple is trying to make sure RLUSD issuance id accurately represented and the need for clawback but I urge you to at least think of another solution.  This would remove a KEY FEATURE THAT liquidity provider can provide liquidity in single deposit $XRP without having to have any connection to Ripple for RLUSD/XRP AMM Pool or the issuer of any other issued tokenized fiat paired to XRP in an AMM POOL. 

If im understanding this correctly, I’ve seen Neil’s posts in the past about the uniqueness of the ability to single deposit XRP without establishing a trustline with issuer of other asset. 

I think  some serious thought should be given to an alternative approach and the Stablecoin issuer should be the one who should at the least consider other options and not immediately change core design features to assist in Ripple not having some additional collateral outstanding. Stablecoin issuers should conform to the protocol and preserving as many aspects of decentralized usability should be a top priority especially at the protocol level. 

BTW I think it should be considered to change “LP Tokens” to simply LIQUIDITY TOKEN as it would resonate far better with XRP Community and better accentuates the fact that LP Tokens are inherently TOKENIZED LIQUIDITY. I think this is especially relevant for XRPL considering the redemption is guaranteed by underlying blockchain native protocol making it a new type premium collateral that is effectively a ~50/50 hybrid stablecoin that can be instantly converted into 3 different ratios of the underlying assets 100% XRP or 100% RLUSD or 50/50 of both. XRP/RLUSD AMM Pool. The better understood XRP/RLUSD LP Tokens are the more likely they will be utilized by owners and DeFi Dapps.

The reality is that Liquid staking tokens have been a far bigger catalyst for the growth and innovative DeFi ecosystem of projects developed around them on ethereum and other networks. 
LIQUIDITY TOKENS are more powerful due to the fact they extend the ability for liquidity providers to harness an arguably superior asset in some use cases than XRP and RLUSD due to the native accrual of yield as more transactions occur with the AMM that change a Fee and CAM’s continuously provide additional source of revenue and harvesting volatility for yield from organic consistent demand for liquidity from ODL and Ripple payments Direct and other use caaes that use AMMs and auto bridge but then there’s also the dual DEX for Trading and with a robust surrounding financial ecosystem of derivatives and other defi protocols composability and additional value from RWAms tokenized and innovative new Funds created that earn yield, this CREATES REAL ECONOMIC VALUE ON THE XRP LEDGER. 

**-MICKEY B. FRESH / @MrFreshTime on X
“The DeFi Standard “**

**Reply by yinyiqian1** [reply]
We are not forcing the liquidity provider to establish a Trustline with the issuer of the other asset in an amm pool. If the provider does not have a trustline with the other asset, the provider can still do single deposit and withdraw.
Here when the provider is not authorized to hold one of the token in the AMM pool, we are proposing that we will prohibit the provider from single depositing the other asset. This happens only when it comes to `authorized trustlines` when the issuer of the token sets `lsfRequireAuth` and the provider here is not authorized to hold that token.
This behaviour also aligns with the proposal that we prohibit single depositing a token when the paired token is frozen.

**Reply by nhartner** [reply]
What happens if the LP provider does a single-sided deposit of asset FOO into a FOO/BAR pool, and AMMClawback is done by FOO issuer. If the LP provider does not have a trust line to BAR issuer, how does the clawback return BAR tokens to the LP provider? Would you force a trust line for BAR to be created on their account?

**Reply by yinyiqian1** [reply]
Yes, a trustline will be create for BAR

**Reply by mvadari** [reply]
Would the user be charged a reserve for that trustline? What if they don't have the XRP for that?

Why not, in this situation, just withdraw all the liquidity in FOO?

**Reply by joynancy** [reply]
If a user holds LP tokens representing FOO/BAR 1:1 in an AMM pool, and the FOO issuer initiates a clawback of all of the user's FOO from AMMs: 

**Current Design (A):**

50% of the user's value, represented in FOO, is clawed back by the issuer. The remaining 50% in BAR is returned to the user, even if no trust line had been previously established for BAR (setting aside the XRP reserve requirement for now). 

**Alternative Design (B):**

Because of user has no trust line established for BAR, and system to avoid establishing a new trust line on-behalf, in this case, when AMMClawback, a single-sided clawback on FOO is initiated, resulting in 100% of the user's value being clawed back by the issuer, leading to a double loss for the user.

In comparison, other LP token holders with both trust lines face only a 50% clawback of their value.  

From my perspective, adopting Design B offers no clear benefit. It significantly disadvantages the user, penalizing them more than necessary. Moreover, I believe issuer is not meant to reclaim more than the 50% value held in FOO by the user. If I were the user being clawback, if 10 XRP reserve is needed, I'd rather pay to keep my 50% value of BAR.

*******
Separately for the XRP reserve, I recall from my experience trading on DEX that even if I swap into an asset for which I don’t have an established trust line, I can still receive the new token. However, in order to use the token, I need to approve the trust line and pay the XRP reserve. This implies that to fully activate a trust line, the user must manually approve it and cover the XRP reserve fee. However, my observation is subject to validation.

**Reply by mvadari** [reply]
* What if BAR is an authorized token, and the issuer hasn't authorized the user to hold the token?
* What if the user doesn't want to hold BAR (perhaps for compliance or tax reasons)?

> Separately for the XRP reserve, I recall from my experience trading on DEX that even if I swap into an asset for which I don’t have an established trust line, I can still receive the new token. However, in order to use the token, I need to approve the trust line and pay the XRP reserve. This implies that to fully activate a trust line, the user must manually approve it and cover the XRP reserve fee. However, my observation is subject to validation.

Yes, but the user essentially signed up to hold that asset by placing the order. They did no such thing when depositing into an AMM - they only signed up to hold the LP Token.

**Reply by joynancy** [reply]
> What if BAR is an authorized token, and the issuer hasn't authorized the user to hold the token?

Refer to 1.2 in the specs: Prevent a wallet from depositing new tokens (either single-sided or double-sided) into an AMM pool if it is not authorized to hold one or both of the tokens in the pool.
Once the change is implemented, I think such scenario shouldn't happen. 

> What if the user doesn't want to hold BAR (perhaps for compliance or tax reasons)?

AMMClawback is essentially an edge case, typically applied when the account holder has been identified as involved in criminal activity or is subject to sanctions. The process of clawback is inherently unpleasant for the affected user, and it is unlikely that any solution will fully satisfy the user. The purpose of creating this primitive is to provide issuers with a compliance tool, allowing them to take whatever actions are necessary to protect the integrity of the financial system. At the same time, the blockchain itself should remain neutral, acting as a platform for enforcement without taking a position on the decisions made by the issuer.

**Comment by yinyiqian1** [comment]
Since we had more discussions, I'll check it again:

Per the new [XLS Contributing process](https://github.com/XRPLF/XRPL-Standards/blob/master/CONTRIBUTING.md), I propose that we can move this discussion to a draft spec: https://github.com/XRPLF/XRPL-Standards/pull/233

Please comment here if you would like to object to moving this spec/discussion forward in the process into a DRAFT spec.
